### PR TITLE
feat: image uploads for notebooks

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

Adds image uploads

### CODE REVIEW

File: frontend/src/scenes/notebooks/Notebook/Editor.tsx [hunk 1]
```diff
@@ -126,14 +130,42 @@ export function Editor({
...
+                        }
+
+                        editor
+                            .chain()
+                            .focus()
+                            .setTextSelection(coordinates.pos)
+                            .insertContent({
+                                type: NotebookNodeType.Image,
+                                attrs: {
+                                    file,
```
Comment (daibhin) [inline frontend/src/scenes/notebooks/Notebook/Editor.tsx:161]: Is this protected from autosaving the file in the contents elsewhere? It looked like that didn't happen based on a quick test but feels like something we shouldn't allow.

File: frontend/src/scenes/notebooks/Nodes/NotebookNodeImage.tsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    return media
+}
+
+const Component = (props: NodeViewProps): JSX.Element => {
+    const { file, src, height } = props.node.attrs
+    const [uploading, setUploading] = useState(false)
+    const [error, setError] = useState<string>()
+
+    useEffect(() => {
```
Comment (daibhin) [inline frontend/src/scenes/notebooks/Nodes/NotebookNodeImage.tsx:32]: Is there a reason this has to be done in the component? Feels like we could remove the `file` attribute from the component if we uploaded the image and retrieved the `src` before ever creating a node

**Review by daibhin — COMMENTED** [review]
Couple of things based on a quick play with images:

- It looks like gifs don't play. Should we limit to png/jpeg while we make sure every file type works?
- When uploading an image it seems to have a lot of whitespace on the top and bottom. Seems to be caused by the `e.currentTarget.naturalHeight` setting of the height. I wonder if we need a better ratio / window width based approach to calculating the height. I know from experience image sizing gets complicated quick so maybe it's fine for now to let people resize
Screenshot 2023-08-07 at 12 16 11
- Resizing the last image is very jumpy. It seems to affect the scroll position and just moves super quickly to the point that you can't control what size it is 
resize-last
- Even if an image is correctly sized on one screen size it will be different on another if purely based on height. I've an (albeit exaggerated) example below of how the same image that fits well on a wide screen actually has a ton of vertical whitespace on a narrow screen.
Screenshot 2023-08-07 at 12 03 13
Screenshot 2023-08-07 at 12 03 24

Most of the things I noticed have to do with resizing. I wonder if there's a better approach to this where we don't add constraints beyond min / max width and heights. Or maybe we could use the settings controls introduced in https://github.com/PostHog/posthog/pull/16910 to allow for a different type of control e.g. force a width / height ratio and let the user choose the object-fit and maybe whether a lightbox is available

**Comment by pauldambra** [comment]
> It looks like gifs don't play. Should we limit to png/jpeg while we make sure every file type works?

https://github.com/PostHog/posthog/pull/16938

Fixed here, silly computers

File: frontend/src/scenes/notebooks/Notebook/Editor.tsx (hunk 1 above)
Comment (benjackwhite) [reply frontend/src/scenes/notebooks/Notebook/Editor.tsx:161]: It _can_ save the contents but they just get saved as an empty object. It felt safe and much more manageable than _not_ including it as part of the state.

File: frontend/src/scenes/notebooks/Nodes/NotebookNodeImage.tsx (hunk 2 above)
Comment (benjackwhite) [reply frontend/src/scenes/notebooks/Nodes/NotebookNodeImage.tsx:32]: In the future this allows us to manage the "unload" state stopping you from navigating away before the file is finished uploading as well as retries etc.

**Comment by benjackwhite** [comment]
@daibhin its hard to reply inline to a review comment, so best to comment on a line or a file even if it is unrelated so that we can inline replies 👍 

> Couple of things based on a quick play with images:
> 
> * It looks like gifs don't play. Should we limit to png/jpeg while we make sure every file type works?

@pauldambra will fix that

> * When uploading an image it seems to have a lot of whitespace on the top and bottom. Seems to be caused by the `e.currentTarget.naturalHeight` setting of the height. I wonder if we need a better ratio / window width based approach to calculating the height. I know from experience image sizing gets complicated quick so maybe it's fine for now to let people resize

Agreed but I think that's something to follow up with. We probably want a "fit to size" type option but I'd rather not over complicate this. Lets see if anyone ever bothers to upload an image and then go from there.

> * Resizing the last image is very jumpy. It seems to affect the scroll position and just moves super quickly to the point that you can't control what size it is

Yeah this applies to all resizing. The hacky fix is to just add a tonne of whitespace at the bottom of the document which means the internal scrolling never kicks in... Don't love it but could be a good follow up.

> * Even if an image is correctly sized on one screen size it will be different on another if purely based on height. I've an (albeit exaggerated) example below of how the same image that fits well on a wide screen actually has a ton of vertical whitespace on a narrow screen.

As said, I vote for just accepting this situation and checking out how its used, unless you have a good solution in mind now?

> Most of the things I noticed have to do with resizing. I wonder if there's a better approach to this where we don't add constraints beyond min / max width and heights. Or maybe we could use the settings controls introduced in #16910 to allow for a different type of control e.g. force a width / height ratio and let the user choose the object-fit and maybe whether a lightbox is available

I think this is the right approach - follow up when we have more of the other work in place and see if an obvious improvement stands out.

**Pull request merged**